### PR TITLE
Set the reason for cancellation in TerminateEnd event

### PR DESCRIPTION
**Pull request by miguelruizdev**
Addressing this issue: https://issues.alfresco.com/jira/browse/AAE-189

### CODE REVIEW

File: activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/TaskCancelledImpl.java [hunk 1]
```diff
@@ -23,16 +23,23 @@
 
 public class TaskCancelledImpl extends RuntimeEventImpl<Task, TaskRuntimeEvent.TaskEvents> implements TaskCancelledEvent {
 
+    private String reason;
+
     public TaskCancelledImpl() {
```
Comment (robfrank) [inline activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/TaskCancelledImpl.java]: is this constructor still useful?
if not, remove and make "reason" final

File: activiti-core/activiti-engine/src/main/java/org/activiti/engine/delegate/event/impl/ActivitiEntityEventImpl.java [hunk 2]
```diff
@@ -24,18 +24,30 @@
...
-  @Override
-  public Object getEntity() {
-    return entity;
-  }
+    @Override
+    public String getReason() {
+        return reason;
+    }
+
+    public void setReason(String reason) {
```
Comment (robfrank) [inline activiti-core/activiti-engine/src/main/java/org/activiti/engine/delegate/event/impl/ActivitiEntityEventImpl.java]: is this setter really necessary? can thew reason be passed as constructor param?

File: activiti-core/activiti-engine/src/main/java/org/activiti/engine/delegate/event/impl/ActivitiEventBuilder.java [hunk 3]
```diff
@@ -388,6 +389,7 @@ protected static void populateEventWithCurrentContext(ActivitiEventImpl event) {
         event.setProcessInstanceId(((Task) persistedObject).getProcessInstanceId());
         event.setExecutionId(((Task) persistedObject).getExecutionId());
         event.setProcessDefinitionId(((Task) persistedObject).getProcessDefinitionId());
+        ((ActivitiEntityEventImpl)event).setReason(TerminateEndEventActivityBehavior.createDeleteReason(null));
```
Comment (robfrank) [inline activiti-core/activiti-engine/src/main/java/org/activiti/engine/delegate/event/impl/ActivitiEventBuilder.java]: this is hurting me

File: activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/TaskCancelledImpl.java (hunk 1 above)
Comment (miguelruizdev) [reply activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/TaskCancelledImpl.java]: True, we can get rid of that no-argument constructor

File: activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/TaskCancelledImpl.java [hunk 4]
```diff
@@ -23,16 +23,23 @@
...
     }
 
-    public TaskCancelledImpl(Task entity) {
+    public TaskCancelledImpl(Task entity, String reason) {
```
Comment (robfrank) [inline activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/TaskCancelledImpl.java]: Can we remove the empty constructor and let "reason" to be final?

File: activiti-core/activiti-engine/src/main/java/org/activiti/engine/delegate/event/impl/ActivitiEventBuilder.java (hunk 3 above)
Comment (miguelruizdev) [reply activiti-core/activiti-engine/src/main/java/org/activiti/engine/delegate/event/impl/ActivitiEventBuilder.java]: it hurts for a reason
this should have never been committed, just a leftover from the debugging session I went through 😄

File: activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/TaskCancelledImpl.java (hunk 4 above)
Comment (miguelruizdev) [reply activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/TaskCancelledImpl.java]: true, that constructor is not used anymore

Comment (miguelruizdev) [reply activiti-core/activiti-api-impl/activiti-api-task-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/TaskCancelledImpl.java]: good catch about the reason field being final (it should never change once the particular instance of the object is created)

File: activiti-core/activiti-engine/src/main/java/org/activiti/engine/delegate/event/impl/ActivitiEntityEventImpl.java (hunk 2 above)
Comment (miguelruizdev) [reply activiti-core/activiti-engine/src/main/java/org/activiti/engine/delegate/event/impl/ActivitiEntityEventImpl.java]: Also, commited by mistake and directly related to the above comment

**Pull request merged**